File: frontend/src/api/reports.ts
import { fetchApi } from './auth';

export type ContentType = 'review' | 'comment';
export type ReportStatus = 'pending' | 'dismissed' | 'actioned';

export interface Report {
  id: number;
  reporter_id: number;
  content_type: ContentType;
  content_id: number;
  reason: string;
  status: ReportStatus;
  created_at: string;
  reviewed_at: string | null;
}

export interface ReportListResponse {
  reports: Report[];
  total: number;
  page: number;
  limit: number;
}

export interface ReportStats {
  pending: number;
  dismissed: number;
  actioned: number;
  total: number;
}

export async function createReport(contentType: ContentType, contentId: number, reason: string): Promise<Report> {
  return fetchApi<Report>('/reports', {
    method: 'POST',
    body: JSON.stringify({
      content_type: contentType,
      content_id: contentId,
      reason,
    }),
  });
}

export async function getPendingReports(page: number = 1, limit: number = 20): Promise<ReportListResponse> {
  return fetchApi<ReportListResponse>(`/reports/admin?page=${page}&limit=${limit}`);
}

export async function getReportStats(): Promise<ReportStats> {
  return fetchApi<ReportStats>('/reports/admin/stats');
}

export async function dismissReport(reportId: number): Promise<Report> {
  return fetchApi<Report>(`/reports/admin/${reportId}/dismiss`, {
    method: 'PATCH',
  });
}

export async function actionReport(reportId: number, removeContent: boolean, warnUser: boolean): Promise<Report> {
  return fetchApi<Report>(`/reports/admin/${reportId}/action`, {
    method: 'PATCH',
    body: JSON.stringify({
      remove_content: removeContent,
      warn_user: warnUser,
    }),
  });
}

File: backend/app/routes/reports.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
import math

from app.dependencies import get_db, get_current_user
from app.middleware.rbac import require_roles
from app.models.user import User
from app.schemas.report import (
    ReportCreate, ReportResponse, ReportListResponse,
    ReportActionRequest, ReportStatsResponse
)
from app.services.report import report_service


router = APIRouter(prefix="/reports", tags=["reports"])
admin_required = Depends(require_roles(["admin"]))


@router.post("", response_model=ReportResponse)
async def create_report(
    report_data: ReportCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Report inappropriate content."""
    report = await report_service.create_report(
        db,
        current_user.id,
        report_data.content_type,
        report_data.content_id,
        report_data.reason,
    )
    return ReportResponse.model_validate(report)


@router.get("/admin", response_model=ReportListResponse, dependencies=[admin_required])
async def list_reports(
    db: AsyncSession = Depends(get_db),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    """List pending reports. Admin only."""
    reports, total = await report_service.get_pending_reports(db, page, limit)
    return ReportListResponse(
        reports=[ReportResponse.model_validate(r) for r in reports],
        total=total,
        page=page,
        limit=limit,
    )


@router.get("/admin/stats", response_model=ReportStatsResponse, dependencies=[admin_required])
async def get_report_stats(db: AsyncSession = Depends(get_db)):
    """Get report statistics. Admin only."""
    stats = await report_service.get_report_stats(db)
    return ReportStatsResponse(**stats)


@router.patch("/admin/{report_id}/dismiss", response_model=ReportResponse, dependencies=[admin_required])
async def dismiss_report(
    report_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Dismiss a report. Admin only."""
    report = await report_service.dismiss_report(db, report_id, current_user.id)
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    return ReportResponse.model_validate(report)


@router.patch("/admin/{report_id}/action", response_model=ReportResponse, dependencies=[admin_required])
async def action_report(
    report_id: int,
    action_data: ReportActionRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Take action on a report. Admin only."""
    report = await report_service.action_report(
        db, report_id, current_user.id,
        action_data.remove_content, action_data.warn_user
    )
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    return ReportResponse.model_validate(report)
